### ISSUE TITLE
Adding new Device Driver to Steam VR folders blocks Steam opening with a DistributedCOM error in Windows Event Viewer

### ISSUE THREAD

**Issue by SamRelativeDimensions** [ValveSoftware/openvr#845]
Hi there,

I've recently been rewriting our old (slightly hacky) Steam VR custom device driver.

After placing the driver folder in SteamVR/drivers with all the other drivers, it causes Steam to fail opening. Steam pops up with it's "Verifying Install..." loading window, which swiftly disappears and does nothing else.

After checking the event viewer, it would appear that it's running into a DistributedCOM error with details:

"The application-specific permission settings do not grant Local Activation permission for the COM Server application with CLSID {ID here} and APPID {steam app id here} to the user {user} from address LocalHost (using LRPC) running in the application container Unavailable SID (Unavailable). This security permission can be modified using the Component Services administrative tool."

Any idea why I might be running into this issue?

Thanks

**Comment by JoeLudwig** [comment]
You shouldn't put drivers into the Steam install directory. Call "vrpathreg.exe adddriver <path>" with the path to your driver and SteamVR will load it from wherever it is.

I don't know for sure that will fix your problem, but it will definitely mean not mixing file permissions in Steam's install directory.

**Comment by SamRelativeDimensions** [comment]
Sure thing, so when I call that I assume I'm calling the driver directory root

vrpathreg.exe adddriver "C:\Program Files\Brand\vrdriverdir\"

Does it provide any sort of output message into cmd when called?

**Comment by walrus** [comment]
You can just run `vrpathreg show` after you've added it to check that it's been added to the list, if that's what you're after?

**Comment by SamRelativeDimensions** [comment]
Yes, or even just running the exe via cmd does the same - it appears in that list, but doesn't load/print anything to log when running Steam VR. Am I supposed to be adding the root path or the dll? I'm assuming the root path

**Comment by SamRelativeDimensions** [comment]
Okay, so after trying the path reg again, Steam VR is finally picking up the directory.

However, upon exiting Steam, and trying to relaunch Steam, I'm still getting the distributed COM error, it seems specifically related to my driver, I'm guessing Steam detects Steam VR drivers upon opening as to begin watchdog threads?

**Comment by SamRelativeDimensions** [comment]
Resolved:

My device driver uses a shared memory buffer to share variables between Open VR driver, Device Driver executable (input from windows), and it's utility suite. It turns out that in my Open VR driver where it was hooking into this shared memory buffer during the Watchdog Drivers Init function, there was a simple programming mistake where it was assumed the memory segment had been hooked into (just a pointer) but the pointer hadn't been assigned yet (as it was slightly delayed on another thread).

TLDR; Totally my fault, trying to access a variable on a pointer that was still nullptr in my WatchdogThreadFunction. whoops! (got to love race conditions right!)

**Comment by JoeLudwig** [comment]
Glad you sorted it out!

And I do love a good race condition.
